File: servers/node/store.js
module.exports.store = (function(){

  var cache = {};
  var nextId = 0;

  return {
    save: function(data) {
      var id = ++nextId;
      var todo = {
        id: id,
        title: data.title,
        completed: data.completed
      }

      cache[id] = todo;

      return todo;
    },

    getAll: function(){
      var todos = []
      for(var key in cache) if(cache.hasOwnProperty(key)) {
          todos.push(cache[key]);
      }
      return todos;
    },

    update: function(id, data) {
      if(cache[id]) {
        var todo = {
          id: id,
          title: data.title,
          completed: data.completed
        }

        cache[id] = todo;

        return todo;
      }
    },

    delete: function(id){
      if (cache[id]) {
        delete cache[id]
        return true;
      }
    }
  }

})();

File: servers/ruby/store.rb
class Store

  def initialize
    @cache = {}
    @next_id = 0
  end

  def save(data)
    @next_id += 1
    todo = {
      'id' => @next_id,
      'title' => data['title'],
      'completed' => data['completed']
    }

    @cache[@next_id] = todo

    todo
  end

  def get_all
    @cache.values
  end

  def update(id, data)
    todo = {
      'id' => id,
      'title' => @data['title'],
      'completed' => @data['completed']
    }

    @cache[id] = todo

    todo
  end

  def delete(id)
    !!@cache.delete(id)
  end

end
